Python Virtual Environment Documentation (#639)
hpcugent/vsc_user_docs: mkdocs/docs/HPC/setting_up_python_virtual_environments.md

# Python Virtual Environments (venv's)

## Introduction

> Python virtual environments (venv's) are a way to create isolated Python environments.

Reviewer: ```suggestion
A Python virtual environment (`venv` for short) is a way to create an isolated Python environment.
```

The last "environment" is a bit too enigmatic imho, especially since "environment" also has a totally different outside of Linux, try to rephrase/expand that a bit?

Author: Applied commit and rephrased last 'environments' in [87a30af](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/87a30afc16070d39726aa37139fbdc62b1162432):

> A Python virtual environment (venv for short)
is a tool to create an isolated Python workspace with its own dependencies.

> You can install packages in a venv without affecting the system Python installation.

Reviewer: ```suggestion
You can install additional Python packages in a virtual environment without affecting the system Python installation.
```

It's not just about not affecting the Python installation, but also controlling *where* additional Python packages are installed, in a controlled way (so they don't go into `.local` in your home directory, etc.)

Author: Applied commit and changed all other occurrences of `venv` with `virtual environment`  in [33ac8a2](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/33ac8a20833fe1dcca543de078d03d72a4a89e7f)

Because a normal user cannot install packages globally, 

> venv's are a way to install packages without needing root access.

Reviewer: `root access` -> `administrator privileges`?

Author: Fixed in [87a30af](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/87a30afc16070d39726aa37139fbdc62b1162432)

This is useful when you need to use a package that is not available as a module on the HPC cluster.

> ## Creating a venv

Reviewer: ```suggestion
## Creating a Python virtual environment
```

Author: fixed in [33ac8a2](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/33ac8a20833fe1dcca543de078d03d72a4a89e7f)

> A venv can be created with the following commands:

Reviewer: ```suggestion
A Python virtual environment can be created with the following commands:
```

Author: fixed in [33ac8a2](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/33ac8a20833fe1dcca543de078d03d72a4a89e7f)

```bash
$ python -m venv myenv      # Create a new venv named myenv
$ source myenv/bin/activate # Activate the venv
```

> After activating the venv, you can install packages with `pip`:

Reviewer: I would avoid using `venv`, use `Python virtual environment` instead.

Also explain better what "activating" does, make a separate subsection for that (create, activate, install, use)

Author: I have replaced some occurrences of `venv` with `python virtual environment`. Other occurrences I just changed to `virtual environment`. See [33ac8a2](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/33ac8a20833fe1dcca543de078d03d72a4a89e7f)

```bash
$ pip install example_package1
$ pip install example_package2
```

It is now possible to run Python scripts that use the installed packages in the venv. To deactivate the venv, run:

```bash
$ deactivate
```

!!! note

>     Always prefer to use modules if they are available. 

Reviewer: You need to clarify what "modules" means here, since there's a lot of room for confusion here with "Python modules".

```suggestion
    Always prefer to use Python packages that are centrally installed, which are available via the environment modules interface.
```

Also link to the section on environment modules we have already.

    Modules are compiled and optimized for the HPC cluster, while packages installed with `pip` are not.

>     To check if a package is available as a module, you can use the following command:

Reviewer: Also here, qualify: `Python package`, `environment module`

Author: Fixed in [d337b6d](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/d337b6d4c075165482d9c80a9f190d54e5864a1b)

    ```bash

>     $ module av package_name

Reviewer: We should also mention that some Python packages are installed as *extensions*, which are part of a particular environment module. Classic examples are `numpy`, `scipy`, `pandas`, which are part of the `SciPy-bundle` module.

You can use `module show` on a `SciPy-bundle` module to check which extensions it provides.

(With some work, we can actually also make extensions visible via `module avail`, but that's not trivial, we haven't found time for it yet. @lbarraga Maybe that's something you can help us with in the coming weeks...)

Author: Added info about bundle modules in [b80340f](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/b80340f01448eb04d010a44d05285b28fece3cc4)

    ```

> ## Combining venv's with modules

Reviewer: reword

Author: fixed in [33ac8a2](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/33ac8a20833fe1dcca543de078d03d72a4a89e7f)

> You can combine packages installed in a venv with modules. The following script uses 

Reviewer: qualify `packages` and `modules`

Author: Fixed in [4f1a65d](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/4f1a65d3fe7341745ac8b3844705d58f3e78f6d6)

> pytorch (which is available as a module) and Poutyne (which is not available as a module):

Reviewer: `pytorch` -> `PyTorch` (the `torch` Python package)

Reviewer: We may install `Poutyne` at some point, so at least say `which we assume is not centrally installed`, or use a fictional example Python packages which doesn't exist (and never will)

Author: Both fixed in [0f603ef](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/0f603efaa8cda5010e587b7f49b43be61478fb0b)

```python title="pytorch_poutine.py"
import torch
import poutyne

...
```

We load a PyTorch package as a module and install Poutyne in a venv:

```bash

> $ ml PyTorch/2.1.2-foss-2023a

Reviewer: use `module load` instead of `ml`

Author: Fixed in [411c9d6](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/411c9d69a6b72d0f5b3b656d4a0cc2fc14e462d5)

$ python -m venv myenv
$ source myenv/bin/activate
$ pip install Poutyne
```

> While in the virtual environment, we can run the script without any issues:

Reviewer: People not familiar with Python virtual environment will not know when they're "in" one, clarify that above when explaining the activation aspect

```bash
$ python pytorch_poutine.py
```

> ## Creating a virtual environment on a specific cluster

Reviewer: `for a specific cluster`?

Author: Fixed in [6754ddd](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/6754ddd8252d1742c4f4686ddf13bbb32cb2d3db)

To create a virtual environment for a specific cluster, you need to start an interactive shell on that cluster.
Let's say you want to create a virtual environment on the `skitty` cluster.

```bash

> $ module swap cluster/skitty

Reviewer 2: skitty is one of the oldest cluster, soon it will be phased out. Please use a newer one, like shinx

Author: changed in [b01aedc](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/b01aedc3b77f1d00c4a3df544f66fa67bbf938fb)

$ qsub -I

> ```

Reviewer: This problem occurs because software install for cluster `shinx` (which is running RHEL9) is not compatible with the `gligar07` and `gligar08` login nodes (which run RHEL8)

> ```

Reviewer: @lbarraga Please use a different cluster for now, `donphan` for example.

After some time, a shell will be started on the `shinx` cluster. 
You can now create a virtual environment as described in [the first section](#creating-a-venv).
This virtual environment can be used by jobs running on the `shinx` cluster.

> !!! warning

Reviewer: Make this a subsection it's likely we'll need to link to this specifically from support tickets

Author: Fixed in [2ae0ad7](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/2ae0ad7fe7329635235f2bbf3847243ae5278857)

    Activating a virtual environment (venv) created on a different cluster can cause issues. 
    This happens because the binaries in the venv from cluster A might not work with the CPU architecture of cluster B.
    
    For example, if we create a venv on the skitty cluster,

    ```bash
    $ module swap cluster/skitty
    $ qsub -I

>     $ python -m venv myenv

Reviewer: I would recommend using `$VSC_OS_LOCAL` + `$VSC_ARCH_LOCAL` here, and creating a subdirectory for all venvs:

```
cd project
mkdir venvs
python -m venv vens/${VSC_OS_LOCAL}-${VSC_ARCH_LOCAL}
pip install -r venvs/requirements.txt
```

```
ml show env/vsc/joltik | grep VSC.*LOCAL
```

helper script `set_up_venv.sh`

```
# check if this script is being sourced, if not exit with usage error
source venvs/load_modules.sh
if [ $? -eq 0 ]; then
  . venvs/${VSC_OS_LOCAL}-${VSC_ARCH_LOCAL}/bin/activate
else
  echo "ERROR ..."
fi
```

`load_modules.sh`
```
module load PyTorch/2.1.2-foss-2023a
```

# Python Virtual Environments (venv's)

## Introduction

A Python virtual environment (venv for short)

> is a tool to create an isolated Python workspace with its own dependencies.

Author: I meant packages, but I see how it would be confusing. Removed in [removed 'dependencies' from sentence](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/2bc34af88ea4cd0bbcecaf9b5f6faa373cd0396b)

you can install additional Python packages without affecting the system-wide Python installation.
Because a normal user cannot install packages globally,
using a virtual environment allows you to install packages locally without needing administrator privileges.
This is especially useful when you need to use a package that is not available as a module on the HPC cluster.

## Managing Python Environments

This section will explain how to create, activate, use and deactivate Python virtual environments.

### Creating a Python virtual environment

A Python virtual environment can be created with the following command:

```bash
$ python -m venv myenv      # Create a new virtual environment named myenv
```

This command creates a new directory named `myenv` in the current working directory. 
This directory will contain the packages, scripts, and binaries that are needed to manage the virtual environment.

### Activating a virtual environment

To use the virtual environment, you need to activate it. 
This will modify the shell environment to use the Python interpreter and packages from the virtual environment.

```bash
$ source myenv/bin/activate # Activate the virtual environment
```

> ### Installing packages in a virtual environment

Author: done in [add section on using venv](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/f3b8a118035fc14809396e2c5d1e27a8d8c03fe1)

    To check if a package is available as a module, you can use the following command:

    ```bash
    $ module av package_name
    ```

    Some Python packages are installed as extensions of modules. For example, `numpy`, `scipy` and `pandas` are 
    part of the `SciPy-bundle` module. You can use

    ```bash
    $ module show <module_name>
    ```

    to check which extensions are included in a module, if any.

### Deactivating a virtual environment

When you are done using the virtual environment, you can deactivate it.
To do that, run:

```bash
$ deactivate
```

## Combining virtual environments with centrally installed modules

You can combine Python packages installed in a virtual environment with environment modules. 
The following script uses PyTorch (which is available as a module)
and Poutyne (which we assume is not centrally installed):

```python title="pytorch_poutyne.py"
import torch
import poutyne

...
```

We load a PyTorch package as a module and install Poutyne in a virtual environment:

```bash
$ module load PyTorch/2.1.2-foss-2023a
$ python -m venv myenv
$ source myenv/bin/activate

> $ pip install Poutyne

Reviewer: `deactivate` at the end here

Author: Fixed in [add deactivate to the end of the Poutyne example](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/66cf0bf20941d392b3c7bc06aa3e474a8d1dd99a)

```

While the virtual environment is activated, we can run the script without any issues:

```bash

> $ python pytorch_poutyne.py

Reviewer: activate first

Author: The environment is activated in the previous codeblock

Reviewer: Yes, but as mentioned elsewhere, I would things apart in an "install" subsection, and a "use" subsection.

And to avoid confusion, it's better to use "`deactivate`" after installing the virtual environment, and `activate` again right before using it (and also load the necessary modules first)

```

## Creating a virtual environment for a specific cluster

To create a virtual environment for a specific cluster, you need to start an interactive shell on that cluster.
Let's say you want to create a virtual environment on the `donphan` cluster.

```bash
$ module swap cluster/donphan
$ qsub -I
```

After some time, a shell will be started on the `donphan` cluster. 
You can now create a virtual environment as described in [the first section](#creating-a-venv).

> This virtual environment can be used by jobs running on the `donphan` cluster.

Reviewer: Maybe already recommend here to use the `$VSC_OS_LOCAL` and `$VSC_ARCH_LOCAL` environment variables to make a cluster-specific virtual environment.

No need to go into details, just mention this:

```shell
python -m venv myenv-${VSC_OS_LOCAL}-{$VSC_ARCH_LOCAL}
source myenv-${VSC_OS_LOCAL}-{$VSC_ARCH_LOCAL}/bin/activate
```

Maybe you can also mention `${VSC_INSTITUTE_CLUSTER}` here, which holds the cluster name.
I doubt many people use all the clusters, they often only use a subset, and the only overlap we have currently is between `donphan` and `joltik`, so this is easier to use than `${VSC_OS_LOCAL}-${VSC_ARCH_LOCAL}` perhaps

Author: Fixed in [add note introducing $VSC_INSTITUTE_CLUSTER](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/6bd51a70e7e2a66bb3c8e679f52b3c532194f3b3). In my opinion, the second option fit the text best

## Example Python job

This section will combine the concepts discussed in the previous sections to:

1. Create a virtual environment on a specific cluster.
2. Combine packages installed in the virtual environment with modules.
3. Submit a job script that uses the virtual environment.

The example script that we will run is the following:

```python title="pytorch_poutyne.py"
import torch
import poutyne

print(f"The version of PyTorch is: {torch.__version__}")
print(f"The version of Poutyne is: {poutyne.__version__}")
```

First, we create a virtual environment on the `donphan` cluster:

```bash
$ module swap cluster/donphan
$ qsub -I
# Load module dependencies
$ module load PyTorch/2.1.2-foss-2023a
$ python -m venv myenv
$ source myenv/bin/activate
# install virtual environment dependencies
$ pip install Poutyne
$ deactivate
```

> We exit the interactive shell by pressing `CTRL+D` and create a job script that loads the PyTorch module, 

Reviewer: use `exit` command instead of `Ctrl-D`

Author: Fixed in ['exit' instead of CTRL+D to exit the interactive shell](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/598ce4f5bfe0dcb3bc3b41e307ce93eaa40ae267)

enters the virtual environment and executes the script:

```bash title="jobscript.pbs"
#!/bin/bash

# Basic parameters
#PBS -N python_job_example    ## Job name
#PBS -l nodes=1:ppn=1         ## 1 node, 1 processors per node
#PBS -l walltime=01:00:00     ## Max time your job will run (no more than 72:00:00)

module load PyTorch/2.1.2-foss-2023a   # Load the PyTorch module

cd $PBS_O_WORKDIR             # Change working directory to the location where the job was submitted

> source myenv/bin/activate     # Activate the virtual environment

Reviewer: move this right below the `module load`, since it's part of setting up the session environment

Author: Done in [source line is now below module load](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/9931d45fde92a052980b74de79ecb1271de1656d)

python pytorch_poutyne.py     # Run your Python script, or any other command within the virtual environment

> deactivate                    # Deactivate the virtual environment

Reviewer: add a blank line above, and maybe mention this is optional (if the job ends, the session disappears, so who cares about activating, but I guess it can help to make things clear)

Author: [add blank line above deactivate](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/a5001ee3343dd4090de5a98faa9b35bf48a41258)

```

Next, we submit the job script:

```bash
$ qsub jobscript.pbs
```

> Two files will be created in the directory where the job was submitted: `python_job_example.o[job_id]` and `python_job_example.e[job_id]`.

Reviewer: Use an example job ID like `123456789`, and say something like "where 123456789 is the job ID of our job"

Author: [use {{jobid}} placeholder](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/a833b02430c595ff47dac252f8261a025347c608)

The `.o` file contains the output of the job.

## Troubleshooting

> ### Illegal instruction error

Reviewer: Make a similar subsection for GLIBC errors which you'll get when running a venv across different major OS versions?

Author: When playing around with this I noticed this is not only the case with venvs, but with almost any command. The following also fails:

```bash
$ ml Python
$ ml swap cluster/shinx
$ ml swap cluster/doduo
/usr/bin/lua: /lib64/libc.so.6: version `GLIBC_2.33' not found (required by /apps/gent/RHEL9/zen4-ib/software/ncurses/6.5-GCCcore-13.3.0/lib/libncurses.so.6)
```

Author: This issue might be worth adding to the general troubleshooting page.

Author: Added in [add GLIBC troubleshooting section](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/999b560119911946e3f2331fbfe60dc58e1fac19).

Reviewer: > This issue might also be worth adding to the general troubleshooting page.

Yes, agreed, there's an entry in your TODO list for this

Activating a virtual environment created on a different cluster can cause issues. 
This happens
because the binaries in the virtual environments from cluster A might not work with the CPU architecture of cluster B.

For example, if we create a virtual environment on the skitty cluster,

```bash
$ module swap cluster/skitty
$ qsub -I
$ python -m venv myenv
```

return to the login node by pressing CTRL+D and try to use the virtual environment:

```bash
$ source myenv/bin/activate
$ python
Illegal instruction (core dumped)
```

we are presented with the illegal instruction error. More info on this [here](troubleshooting.md#illegal-instruction-error)

### Error: cannot open shared object file: No such file or directory

There are two main reasons why this error could occur.

> 1. you have not loaded the python module that was used to create the virtual environment.

Reviewer: ```suggestion
1. You have not loaded the `Python` module that was used to create the virtual environment.

> 2. you added or removed modules while the virtual environment was activated.

Reviewer: ```suggestion
2. You added or removed modules while the virtual environment was activated.

> #### Entering a virtual environment while the python module used to create it is not active

Reviewer: Use "`Python` module", same further below

```suggestion
#### Entering a virtual environment while the Python module used to create it is not active

> When you load a python module and use that to make a virtual environment, you need to make sure that the same module 

Reviewer: ```suggestion
If you loaded a `Python` module when creating a virtual environment, you need to make sure that the same module

is loaded when you enter the environment. This is because the virtual environment keeps a reference to the base python 
used to create it.

The following commands illustrate this issue:

```bash
$ module load Python/3.10.8-GCCcore-12.2.0  # Load a python module
$ python -m venv myenv                      # Create a virtual environment with loaded python module

> $ module purge                              # Remove all loaded modules

Reviewer: mention `(WRONG!)` or something in the comment to highlight that this is where things go wrong, and maybe use `module unload Python` instead to make it a bit more explicit?

Mentioning `module purge` is good too though, since that may be more common than selective unloading modules

Author: Fixed in [clarify wrong command in example](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/1a2f3f1a2cb4041e239bbaad9a3b454d6b084c13)

$ source myenv/bin/activate                 # Activate the virtual environment
$ python                                    # Start python
python: error while loading shared libraries: libpython3.10.so.1.0: cannot open shared object file: No such file or directory
```

Here, the virtual environment tries to use the python module that was loaded when the environment was created, which is no longer available.
The solution is to load the same python module before activating the virtual environment:

```bash
$ module load Python/3.10.8-GCCcore-12.2.0  # Load the same python module
$ source myenv/bin/activate                 # Activate the virtual environment
```

> #### modifying modules while in a virtual environment

Reviewer: ```suggestion
#### Modifying modules while in a virtual environment
```

But this actually needs a better subtitle (see below), maybe "Loading/unloading modules while in ..."?

You must not delete or add modules while in a virtual environment. 

> Adding and removing modules modifies the `$PATH` variable in the current shell. When activating a virtual environment,

Reviewer: Be consistent in terminology, don't use "delete", "remove", "unload" to all mean the same time, only use "unload"

Same for "add" vs "load", only use "load"

Author: Fixed in [6fa2f76](https://github.com/hpcugent/vsc_user_docs/pull/639/commits/6fa2f76d3cc850a43bd5f8c9d8462d0b4f9c5a5e)
